sympy/functions/special/delta_functions.py

```diff
@@ -159,6 +159,36 @@ def is_simple(self, x):
             return p.degree() == 1
         return False
 
+    def _eval_rewrite_as_Piecewise(self, *args):
+        """Represents DiracDelta in a Piecewise form
+
+           Examples
+           ========
+
+           >>> from sympy import DiracDelta, Piecewise, Symbol, oo
+           >>> x = Symbol('x', real=True)
+
+           >>> DiracDelta(x).rewrite(Piecewise)
+           Piecewise((oo, Eq(x, 0)), (0, True))
+
+           >>> DiracDelta(x - 5).rewrite(Piecewise)
+           Piecewise((oo, Eq(x - 5, 0)), (0, True))
+
+           >>> DiracDelta(x - 5, 4).rewrite(Piecewise)
```

Comment:
The DiracDelta docstring really ought to describe what the heck a delta function is, and how it works in SymPy. Mathematically, it's not really a function, it's a distribution (roughly speaking, this means it only makes sense inside an inner product, namely, as `Integral(DiracDelta(x)*f(x), (x, -oo, oo))`, which equals `f(0)`). Its derivative is also not a real derivative, but a distributional derivative, which is roughly defined by taking the integration by parts rule of "moving the derivative" to the other function in the integral. 

Rewriting a delta as a piecewise is already incorrect, but rewriting a delta derivative seems even more incorrect somehow. I think it would be fine to disallow this case. Then again, if someone is rewriting deltas as piecewise, maybe they really want this, so maybe we can leave this as is. 

`deltaintegrate` is the algorithm for integrating delta functions. The usual integration algorithms don't apply because the delta function behaves different from usual functions in integrals (pretty much by definition).